src/core/control/Control.cpp

```diff
@@ -1780,6 +1780,25 @@ void Control::undoRedoPageChanged(PageRef page) {
 }
 
 void Control::selectTool(ToolType type) {
+    // keep text-selection when switching from text to seletion tool
+    auto oldTool = getToolHandler()->getActiveTool();
+    if (oldTool && win
+                && isSelectToolType(type)
+                && oldTool->getToolType() == ToolType::TOOL_TEXT
+                && !(this->win->getXournal()->getTextEditor()->getText()->getText().empty())) {
+        auto xournal = this->win->getXournal();
+        Text* textobj = xournal->getTextEditor()->getText();
+        clearSelectionEndText();
+
+        auto pageNr = getCurrentPageNo();
+        XojPageView* view = xournal->getViewFor(pageNr);
+        g_assert(view != nullptr);
+        PageRef page = this->doc->getPage(pageNr);
+        auto selection = new EditSelection(this->undoRedo, textobj, view, page);
```

Comment:
Wait! Sorry I took a long time to check this out, but...
I fear there may be some "double ownership" cases here (textobj may be owned by both the layer and the selection). Let me check before merging.

Reply:
Ah, no, it seems to be ok. EditSelection(this->undoRedo, textobj, view, page); removes the element from the layer.

We need to lock the document's mutex though
```suggestion
        Text* textobj = xournal->getTextEditor()->getText();
        clearSelectionEndText();

        auto pageNr = getCurrentPageNo();
        XojPageView* view = xournal->getViewFor(pageNr);
        g_assert(view != nullptr);
        this->doc->lock()
        PageRef page = this->doc->getPage(pageNr);
        auto selection = new EditSelection(this->undoRedo, textobj, view, page);
        this->doc->unlock();
```